bokeh/core/serialization.py

```diff
@@ -0,0 +1,298 @@
+        return obj in self._references
+
+    def add_ref(self, obj: Any, ref: Reference) -> None:
+        self._references[obj] = ref
+
+    def del_ref(self, obj: Any) -> None:
+        del self._references[obj]
+
+    def add_buf(self, obj: bytes) -> Buffer:
+        buffer = Buffer(self._next_id(), obj)
+        self._buffers.append(buffer)
+        return buffer
+
+    @property
+    def references(self) -> List[Reference]:
+        return list(self._references.values())
+
+    @property
+    def buffers(self) -> List[Buffer]:
+        return list(self._buffers)
+
+    _id: int = 0
+    def _next_id(self) -> ID:
+        self._id += 1
+        return ID(str(self._id))
+
+    def _encode(self, obj: Any) -> JSON:
+        if isinstance(obj, Serializable):
+            return obj.to_serializable(self)
+        elif type(obj) in _serializers:
+            return _serializers[type(obj)](self)
+        elif obj is None:
+            return None
+        elif isinstance(obj, bool):
+            return self._encode_bool(obj)
+        elif isinstance(obj, str):
+            return self._encode_str(obj)
+        elif isinstance(obj, int):
+            return self._encode_int(obj)
+        elif isinstance(obj, float):
+            return self._encode_float(obj)
+        elif isinstance(obj, tuple):
+            return self._encode_tuple(obj)
+        elif isinstance(obj, list):
+            return self._encode_list(obj)
+        elif isinstance(obj, dict):
+            return self._encode_dict(obj)
+        elif isinstance(obj, np.ndarray):
+            return self._encode_ndarray(obj)
+        else:
+            return self._encode_other(obj)
+
+    def _encode_bool(self, obj: bool) -> JSON:
+        return obj
+
+    def _encode_str(self, obj: str) -> JSON:
+        return obj
+
+    def _encode_int(self, obj: int) -> JSON:
+        if -MAX_INT < obj <= MAX_INT:
+            return obj
+        else:
+            log.warn("out of range integer may result in loss of precision")
+            return self._encode_float(float(obj))
+
+    def _encode_float(self, obj: float) -> JSON:
+        if isnan(obj):
+            return {"$type": "number", "value": "nan"}
+        elif isinf(obj):
+            return {"$type": "number", "value": f"{'-' if obj < 0 else '+'}inf"}
+        else:
+            return obj
+
+    def _encode_tuple(self, obj: Tuple[Any, ...]) -> JSON:
+        return self._encode_list(list(obj))
+
+    def _encode_list(self, obj: List[Any]) -> JSON:
+        return [self._encode(item) for item in obj]
+
+    def _encode_dict(self, obj: Dict[Any, Any]) -> JSON:
+        return {self._encode(key): self._encode(val) for key, val in obj.items()}
+
+    def _encode_ndarray(self, obj: npt.NDArray[Any]) -> JSON:
+        array = transform_array(obj)
+
+        if array_encoding_disabled(array):
+            data = self._encode_list(array.tolist())
+        else:
+            buffer = self.add_buf(array.tobytes())
```

Comment:
we would have to look at what Tornado can support these days, but `array.data` or`bytearray(array)` might be able to avoid the extra copy that `tobytes` makes